[1.19.3] Add onStopUsing hook to IForgeItem

KnightMiner opened the pull request:
## New hook

This PR introduces `IForgeItem#onStopUsing`, which is called in `LivingEntity#stopUsingItem`. This hook runs on both client and serverside after you stop using an item, whether you let go of right click, scroll to a different item, swap items, or are using a shield that is "broken" from an axe


There currently only two known cases the hook does not trigger, both of which are vanilla bugs fixed by #9344:
* When you drop an item, due to [MC-231097](https://bugs.mojang.com/browse/MC-231097).
* On the serverside when a shield breaks in use, due to [MC-168573](https://bugs.mojang.com/browse/MC-168573)

## Justification

Existing hooks (`finishUsingItem` and `releaseUsing`) are only called when you release right click on the item. While this is useful for canceling firing bows, it can prevent you from disabling an effect of using an item that was enabled in `use`. One workaround is to use the vanilla `ITEM_INTERACT_FINISH` game event, but that only runs serverside making it insufficient for things such as FOV changes (via `ComputeFovModifierEvent`) or some player abilities. Another workaround is the `LivingEquipmentChangeEvent`, which is much less reliable for this usecase and again, serverside only.

## Test Mod

Included is a test mod that adds two items: `stop_using_item:bad_scope` and `stop_using_item:good_scope`. Both items when activated grant you creative flight and increase the FOV by a significant amount. With `bad_scope`, you can observe that using the scroll wheel does not reset the FOV. 

## Vanilla Bug

As mentioned earlier, a vanilla bug causes you to keep using an item after dropping it or when a shield breaks. I implemented those fixes in #9344. The two PRs can be merged separately, but it will make this hook much nicer to use if the other PR is merged.

KnightMiner committed:
Add onStopUsing hook to IForgeItem

Called when you stop using an item for any reason, baring MC-231097

KnightMiner committed:
Remove javadoc mentioning a vanilla bug fixed by my other PR

KnightMiner committed:
Added license to new file

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/item/StopUsingItemTest.java:
```diff
@@ -0,0 +1,144 @@
+import net.minecraftforge.registries.RegistryObject;
+
+@Mod(StopUsingItemTest.MODID)
+public class StopUsingItemTest
```
Please provide a javadoc for this class which explains what feature this tests, how to test it, and what the expected results are from said test.

sciwhiz12 reviewed (changes requested):
Works fine in testing, just have one quibble with regards to test mod documentation.

KnightMiner committed:
Add javadoc for test

LexManos commented on patches/minecraft/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -600,6 +600,14 @@
+ 
+    public void m_5810_() {
++      if (this.m_6117_() && !this.f_20935_.m_41619_()) this.f_20935_.onStopUsing(this, f_20936_);
+       if (!this.f_19853_.f_46443_) {
```
I think this is missing the use case in Player.hurtCurrentlyUsedShield
The documentation is also a little unclear, it should state that it is called in ADDITION to those events.
To make it clear that both will be fired.
Also, would this be a good time to add a start using item level callback to mirror this?

KnightMiner replied:
I believe Player.hurtCurrentlyUsedShield just calls this method directly, hence me putting the patch in LivingEntity. I'll double check that when I get a chance.

----

I cannot think of any usecase for a start using item callback at the item level, since you start using an item by calling a method in your own item code. Typically that is just in a single method as well, `Item#use`, if you need to do it in multiple places you can always add your own function to call in the mod instead of needing a forge change.

Doing it at the event level might be reasonable (with an event for stop using as well), but I feel that would be better left to a future PR if desired as the usecases are quite different (and personally, I cannot think of a good usecase for detecting when another mod starts using their item)

KnightMiner replied:
Had a chance to look at the code, I believe `hurtCurrentlyUsedShield` might be the same sort of case as [MC-231097](https://bugs.mojang.com/browse/MC-231097) (closest I can find is [MC-168573](https://bugs.mojang.com/browse/MC-168573)) where you actually never stop using the item when its broken. I could easily add a patch into that method to call `stopUsingItem` when the shield breaks to fix that though, hard to say whether that is in scope as its another vanilla bug to fix (arguably belongs in #9344)

KnightMiner replied:
Fix for the shield issue has been added to #9344, after confirmation that it is in fact reported to vanilla as MC-168573

KnightMiner committed:
Note that the new hook is called in addition to other hooks, running after when both are run

KnightMiner committed:
Added a note to the javadoc about a potential issue with breaking stacks

With vanilla shields, that is addressed by a Forge fix now, but if a modded item simiarly damages the item while in use (e.g. in onUsingTick), the same issue may come up again.
